### PR TITLE
add `OpInfo`s for `torch.nn.functional.triplet_margin(_with_distance)?_loss`

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65681
* #68419
* #69212
* __->__ #67079
* #67023
* #67022
* #69211

cc @albanD @mruberry @jbschlosser @walterddr @kshitij12345

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -14341,7 +14362,37 @@ def ref_pairwise_distance(input1, input2):
...
+        supports_out=False,
+    ),
+    OpInfo(
+        "nn.functional.triplet_margin_with_distance_loss",
+        sample_inputs_func=partial(sample_inputs_triplet_margin_loss, with_distance=True),
+        dtypes=all_types_and_complex_and(torch.bfloat16),
+        dtypesIfCUDA=all_types_and_complex_and(torch.float16, torch.bfloat16),
+        supports_out=False,
+        skips=(
+            # This test cannot handle a callable passed to `distance_function`. If we would use
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:14380]: Nice xfails

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -7593,6 +7594,26 @@ def sample_inputs_binary_cross_entropy(op_info, device, dtype, requires_grad, lo
...
+        dict(swap=True),
+        *[dict(reduction=reduction) for reduction in ("mean", "sum", "none")],
+    )
+
+    sample_inputs = []
+    for kwargs in kwargss:
+        input = make()
+        args = (make(), make())
+        if with_distance:
+            kwargs["distance_function"] = nn.PairwiseDistance(p=kwargs.pop("p", 2.0), eps=kwargs.pop("eps", 1e-6))
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7612]: Let's make this more fun by using cosine similarity:

```
lambda x, y: 1.0 - F.cosine_similarity(x, y)
```

Comment (pmeier) [reply torch/testing/_internal/common_methods_invocations.py]: I've made this change, but I'm going to revert it, since `F.cosine_similarity` only supports a subset of the dtypes of the op. See https://app.circleci.com/pipelines/github/pytorch/pytorch/419531/workflows/90e12458-df8f-4383-b995-9bcfd53157da/jobs/16818402 for some failures. Only other remedy would be to limit the supported dtypes of the `OpInfo` to the subset of the loss functional.

**Review by mruberry — APPROVED** [review]

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py]: Thanks @pmeier

**Comment by pmeier** [comment]
Superseded by #75782.
